dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,203 @@
+with the connected Fobnail Token, should allow determining the state of the
+platform before the target OS is launched. As we can see in the presentation,
+`heads` would be responsible for the platform's trust control, while in this
+document, we want to present the results of research on the possibilities of
+using various operating systems.
+
+## Different OSs propositions
+
+The research effect is presented below. 4 systems were considered:
+
+* Zephyr
+* Xous
+* seL4
+* Linux
+
+Each of them has a short description, an analysis of the launch in DLME, and the
+possibilities and potential problems that will have to be addressed for the
+Fobnail Token to be functional.
+
+### Zephyr
+
+Zephyr is a well-known and supported embedded RTOS with a monolithic kernel
+written in C. It supports various devices ranging from Cortex-M to x86 CPUs and
+provides support for IPv4, CoAP, and USB needed for communication with Fobnail
+Token.
+
+#### Running in DLME
+
+Zephyr boots over Multiboot1, but SKL supports only Muliboot2, so either SKL
+must be extended to gain Multiboot1 support or Zephyr to gain Multiboot2. Also,
+Zephyr is not portable - some important configuration is baked into Zephyr
+during build, like LAPIC base address - the base address is set by
+`CONFIG_LOAPIC_BASE_ADDRESS`, if `IA32_APIC_BASE` MSR doesn't match, Zephyr will
+break. The same thing for xAPIC vs x2APIC mode. If Zephyr was configured to use
+xAPIC mode, it might not boot on a platform with x2APIC enabled in its firmware.
+
+#### Fobnail integration
+
+Zephyr provides most of the drivers needed for integration:
+
+- USB hub drivers
+- USB EEM driver
+- Network stack with IPv4 and CoAP support
+
+TPM driver is missing, however there is a fairly recent
+[PoC implementation](https://github.com/drandreas/zephyr-tpm2-poc) of TPM2
+stack.
+
+Zephyr provides good support for standard C library (newlib), which should
+simplify [fobnail-attester](https://github.com/fobnail/fobnail-attester)
+porting.
+
+### Xous
+
+Xous is a microkernel-based OS written in Rust, currently in Alpha state. Xous
+is used as a secure microkernel for [Betrusted](betrusted.io) device and is
+focused on RISC-V, which right now is the only architecture supported. However,
+x86 already contains
+[dummy implementation](https://github.com/betrusted-io/xous-core/blob/30b82b25b100e958790973c129dc49e1acca79ec/kernel/src/arch/x86_64.rs)
+which probably will be extended someday.
+
+
+#### Running in DLME
+
+Xous currently cannot run on x86, and DLME on RISC-V is not covered here.
+
+#### Fobnail integration
+
+Xous has network support but currently lacks USB host drivers and TPM support.
+Also, it lacks C interface for calling the kernel and lacks C library. Getting
+[fobnail-attester](https://github.com/fobnail/fobnail-attester) running may
+require a significant amount of work. Also, Xous needs to gain kexec-like
+abilities to chainload target OS.
+
+### seL4
```

Comment:
Please reach seL4 community and let them know we consider use of the project, maybe some information can be improved here.

Reply:
Submitted https://github.com/seL4/seL4/issues/832

Reply:
submitted and added to the report

Reply:
Thanks!